app/controllers/teachers_controller.rb

```diff
@@ -18,14 +18,18 @@ class TeachersController < ApplicationController
   end
 
   post '/teachers/signup' do
-    @teacher=Teacher.create(first_name: params[:first_name], last_name: params[:last_name], preferred_name: params[:preferred_name], password: params[:password])
-    @teacher.save
+    @teacher = Teacher.new(first_name: params[:first_name], last_name: params[:last_name], preferred_name: params[:preferred_name], username: params[:username], password: params[:password])
+    if @teacher.save
     flash[:message]="You have successfully created an account. Now you can log in!"
     redirect to '/teachers/login'
+    else
+      flash[:message]="That username is already taken. Try another one."
+      redirect 'teachers/signup'
+    end
   end
 
   post '/teachers/login' do
-    @teacher=Teacher.find_by(first_name: params[:first_name], last_name: params[:last_name])
+    @teacher=Teacher.find_by(username: params[:username])
```

Comment:
this route should be authenticating a teacher

Reply:
isn't that what happens in the next line or do you want to see authenticate_teacher! method?